src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,331 @@
+package consistenthashing
+
+import (
+	"context"
+	"hash/crc32"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"go.etcd.io/etcd/client/v3/concurrency"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/gogo/protobuf/types"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dlock"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+
+	"github.com/sirupsen/logrus"
+
+	etcd "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	// these variables are overridable for testing purposes.
+	hashFn = crc32.ChecksumIEEE
+	idGen  = uuid.New
+)
+
+// Ring is a consistent hash Ring. Each process should only create one instance with a given prefix. Nodes can be
+// added locally to a Ring. A Ring watches for changes to the prefix to determine if new nodes have been added
+// or deleted.
+type Ring struct {
+	client    *etcd.Client
+	hashFn    func([]byte) uint32
+	stateLock sync.Mutex
+	logger    *logrus.Logger
+	members   []member
+	node      node // node represents the member instance in a Ring that is local to the current process.
+	prefix    string
+}
+
+// node is a local member of a Ring. Upon creation, a node requests a lease from etcd and continuously refreshes it
+// until it is shutdown.
+type node struct {
+	member
+	locks map[string]lockInfo
+}
+
+// member contains metadata about members in a Ring.
+type member struct {
+	Id   string
```

Comment:
unexport

Reply:
We want this exported so we can return information about Members()